timemachine/fe/dummy.py

```diff
@@ -1,151 +1,202 @@
-    sparse_graph = []
-    for row in dense_graph:
-        nbs = []
-        for col_idx, col in enumerate(row):
-            if col == 1:
-                nbs.append(col_idx)
-        sparse_graph.append(nbs)
-
-    # conditional depth first search that terminates when
-    # we encounter a core atom.
-    def conditional_dfs(i, visited):
-        if i in visited:
-            return
-        else:
-            visited.add(i)
-            if i not in core:
-                for nb in sparse_graph[i]:
-                    conditional_dfs(nb, visited)
+import networkx as nx
 
-    visited = set()
+from timemachine.fe.utils import get_romol_bonds
 
-    conditional_dfs(dummy_atom, visited)
 
-    anchors = [a_idx for a_idx in visited if a_idx in core]
+class MultipleAnchorWarning(UserWarning):
+    pass
 
-    return anchors
 
+def generate_dummy_group_assignments(
+    bond_graph: nx.Graph, core_atoms: Collection[int]
+) -> Iterator[Dict[int, FrozenSet[int]]]:
+    """Returns an iterator over dummy group assignments (i.e., candidate partitionings of dummy atoms with each
+    partition assigned a bond anchor atom) for a given molecule (represented as a bond graph) and set of core atoms.
 
-def identify_dummy_groups(bond_idxs, core):
-    """
-    A dummy group is a set of dummy atoms that are inserted or deleted in alchemical
-    free energy calculations. The bonded terms that involve dummy atoms need to be
-    judiciously pruned such that the partition function at the end-states remain
-    factorizable, and cancelleable. Dummy groups are subject to the following constraints:
+    A dummy group is a set of dummy atoms that are inserted or deleted in alchemical free energy calculations. The
+    bonded terms that involve dummy atoms need to be judiciously pruned such that the partition function at the
+    end-states remain factorizable, and cancelleable. Dummy groups are subject to the following constraints:
 
     1) They must not contain atoms in the core.
     2) Dummy groups do not interact with other dummy groups.
     3) Dummy groups interact with the core only through anchor atoms.
 
-    While the choices of dummy groups is arbitrary, this function partitions the dummy
-    atoms into multiple dummy groups using a heuristic:
-
-    1) Identify all 1-2, 1-3 edges between dummy atoms.
-    2) Generate an induced graph using edges from above set.
-    3) Disconnected components of the resulting induced graph are our dummy groups.
+    While the choices of dummy groups is arbitrary, this function partitions the dummy atoms into multiple dummy groups
+    using a heuristic:
 
-    Code example:
+    1) Generate an induced subgraph containing only dummy atoms
+    2) Identify connected components of the induced subgraph
+    3) For each connected component, generate all possible choices of (bond) anchor atom
+    4) Merge connected components with the same bond anchor; this gives a mapping of anchor atom to dummy group for each
```

Comment:
missing a step: between 3) and 4) - the choice of the bond anchor that a dummy group belongs to is currently done randomly?

Reply:
It is not done randomly; the returned iterator will generate the outcomes for all possible choices. Will try to clarify here.

Reply:
Attempted to clarify in https://github.com/proteneer/timemachine/pull/1019/commits/daf51795268dac55e11595665deb7d51e2fd8566